File: LeetCode/0097.交错字符串.cpp
/*
 * @lc app=leetcode.cn id=97 lang=cpp
 *
 * [97] 交错字符串
 */
#include <string>
using namespace std;
// @lc code=start
// 这个方法最后几个测试点会超时，但是总的来说可以使用
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        return check(s1, s2, s3, 0, 0, 0);
    }

    bool check(string s1, string s2, string s3, int i, int j, int k) {
        if (k == s3.size()) {
            return i == s1.size() && j == s2.size();
        }
        if (s1[i] == s3[k] && check(s1, s2, s3, i + 1, j, k + 1)) {
            return true;
        }
        if (s2[j] == s3[k] && check(s1, s2, s3, i, j + 1, k + 1)) {
            return true;
        }
        return false;
    }
};
// @lc code=end

File: LeetCode/0097.交错字符串.java
/*
 * @lc app=leetcode.cn id=97 lang=java
 *
 * [97] 交错字符串
 */

// @lc code=start
class Solution {
     public boolean isInterleave(String s1, String s2, String s3) {
        if (s1.length() + s2.length() != s3.length()) return false;

        char[]sub1 = s1.toCharArray();
        char[]sub2 = s2.toCharArray();
        char[] str = s3.toCharArray();

        boolean[][]dp = new boolean[sub1.length + 1][sub2.length + 1];
        dp[0][0] = true;

        for (int i = 0; i <= sub1.length; i++) {
            for (int j = 0; j <= sub2.length; j++) {
                if (i > 0 && sub1[i - 1] == str[i + j - 1] && dp[i - 1][j]) dp[i][j] = true;
                if (j > 0 && sub2[j - 1] == str[i + j - 1] && dp[i][j - 1]) dp[i][j] = true;
            }
        }

        return dp[sub1.length][sub2.length];
    }
}
// @lc code=end

